servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ProxyTunnel.java

```diff
@@ -105,72 +106,65 @@ int connectCount() {
-
-                try (Socket clientSocket = new Socket(host, port)) {
-                    connectCount.incrementAndGet();
-                    final OutputStream out = socket.getOutputStream();
-                    out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
-                    out.flush();
-
-                    final InputStream cin = clientSocket.getInputStream();
-                    Future<Void> f = executor.submit(() -> {
-                        copyStream(out, cin);
-                        return null;
-                    });
-                    copyStream(clientSocket.getOutputStream(), in);
-                    f.get(); // wait for the copy of proxy client input to server output to finish copying.
-                }
-            } else {
-                throw new RuntimeException("Unrecognized initial line: " + initialLine);
+    private void handleRequest(final Socket socket, final String initialLine) throws IOException {
+        if (initialLine.startsWith(CONNECT_PREFIX)) {
+            final int end = initialLine.indexOf(' ', CONNECT_PREFIX.length());
+            final String authority = initialLine.substring(CONNECT_PREFIX.length(), end);
+            final String protocol = initialLine.substring(end + 1);
+            final int colon = authority.indexOf(':');
+            final String host = authority.substring(0, colon);
+            final int port = Integer.parseInt(authority.substring(colon + 1));
+
+            try (Socket clientSocket = new Socket(host, port)) {
+                connectCount.incrementAndGet();
+                final OutputStream out = socket.getOutputStream();
+                out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
+                out.flush();
+
+                executor.submit(() -> {
+                    try {
+                        copyStream(out, clientSocket.getInputStream());
+                    } finally {
+                        // We are simulating a proxy that doesn't do half closure. The proxy should close the server
+                        // socket as soon as the server read is done.
+                        socket.close();
+                    }
+                    return null;
+                });
+                copyStream(clientSocket.getOutputStream(), socket.getInputStream());
             }
-        } finally {
-            in.close();
+        } else {
+            throw new IllegalArgumentException("Unrecognized initial line: " + initialLine);
         }
+        // The Socket is closed outside the scope of this method.
     }
 
     private static void copyStream(final OutputStream out, final InputStream cin) throws IOException {
-        try {
-            int b;
-            while ((b = cin.read()) >= 0) {
-                out.write(b);
-            }
+        int read;
+        final byte[] bytes = new byte[256];
```

Comment:
As described in https://github.com/apple/servicetalk/pull/1817#discussion_r705563782, we need to fragment data as much as possible. Consider keeping it as is (read by one byte) or use a small buffer, around 8 bytes. The overall size of transferred data in tests is small. I don't think it's a perf concern or something that can make CI slower.

Reply:
used 8, and added a comment